File: lincoln/ruby/lib/lang/plan/array_index.rb
class ArrayIndex < Expression
  def initialize(array, index)
		# assert(array.type().isArray());
		@array = array
		@index = index
	end
	
  class type
    return @array.type.getComponentType
  end
	
	def to_s
		return "(" + @array.to_s + ")[" + @index.to_s + "]"
	end

	def variables
		return @array.variables
	end

	def function
	  doit = lambda do |t|
      return @array.function.call(tuple)[index]
    end
    tmpClass = Class.new(TupleFunction)

    tmpClass.send :define_method, :evaluate do |tuple|
      return doit.call(tuple)
    end
    tmpClass.send :define_method, :returnType do 
      return type 
    end
    retval = tmpClass.new
    return retval
	end
end

File: lincoln/java/dataflow/src/jol/lang/plan/ArrayIndex.java
package jol.lang.plan;

import java.lang.reflect.Array;
import java.util.Set;

import jol.types.basic.Tuple;
import jol.types.basic.Schema;
import jol.types.exception.JolRuntimeException;
import jol.types.exception.PlannerException;
import jol.types.function.TupleFunction;

public class ArrayIndex extends Expression {
	
	private Expression<Variable> array;
	
	private Integer index;
	
	public ArrayIndex(Expression<Variable> array, Integer index) {
		assert(array.type().isArray());
		this.array = array;
		this.index = index;
	}
	
	@Override
	public Expression clone() {
		return new ArrayIndex(array, index);
	}
	
	@Override
	public Class type() {
		return array.type().getComponentType();
	}
	
	@Override
	public String toString() {
		return "(" + array.toString() + ")[" + index + "]";
	}

	@Override
	public Set<Variable> variables() {
		return array.variables();
	}

	@Override
	public TupleFunction function(Schema schema) throws PlannerException {
		final TupleFunction arrayFunc = array.function(schema);
		return new TupleFunction() {
			public Object evaluate(Tuple tuple) throws JolRuntimeException {
				return Array.get(arrayFunc.evaluate(tuple), index);
			}

			public Class returnType() { return type(); }
			
		};
	}
}
